Sources/SwiftProtobuf/StringUtils.swift

```diff
@@ -8,32 +8,115 @@
 //
 // -----------------------------------------------------------------------------
 ///
-/// Utility functions that are generally useful...
+/// Utility functions for converting UTF8 bytes into Strings.
+/// These functions must:
+///  * Accept any valid UTF8, including a zero byte (which is
+///    a valid UTF8 encoding of U+0000)
+///  * Return nil for any invalid UTF8
+///  * Be fast (since they're extensively used by all decoders
+///    and even some of the encoders)
 ///
 // -----------------------------------------------------------------------------
 
 import Foundation
 
-// Convert a pointer/count that refers to a block of UTF8 bytes
-// in memory into a String.
-// Returns nil if UTF8 is invalid.
+// Wrapper that takes a buffer and start/end offsets
+internal func utf8ToString(
+  bytes: UnsafeBufferPointer<UInt8>,
+  start: UnsafeBufferPointer<UInt8>.Index,
+  end: UnsafeBufferPointer<UInt8>.Index
+) -> String? {
+  return utf8ToString(bytes: bytes.baseAddress! + start, count: end - start)
+}
+
+// Swift's support for working with UTF8 bytes directly has
+// evolved over time.  The following tries to choose the
+// best option depending on the version of Swift you're using.
+
+#if swift(>=4.0)
+
+///////////////////////////
+//
+// Swift 4 (all platforms)
```

Comment:
Do you want to use `// MARK: - Blah blah` so Xcode helps group them in the jump bar?